File: nodejs/database/shipyards.ts
import { DbData, db } from './db.ts';
import { Shipyard } from '../lib/types';

const addStatement = db.prepare(`INSERT INTO shipyards(data, updated) VALUES (json(:data), :date);`);
const getStatement = db.prepare(`SELECT data FROM shipyards WHERE data->>'symbol' = ?;`);
const updateStatement = db.prepare(`UPDATE shipyards SET data = json(:data), updated = :date WHERE data->>'symbol' = :symbol;`);

export function get(symbol: string): Shipyard|null {
	const data = getStatement.get(symbol) as DbData|undefined;
	if (!data) return null;
	return JSON.parse(data.data);
}

export function set(data: Shipyard): void {
    if (get(data.symbol) === null) {
		addStatement.run({
			data: JSON.stringify(data),
			date: new Date().toISOString(),
		});
    } else {
		updateStatement.run({
			data: JSON.stringify(data),
			date: new Date().toISOString(),
			symbol: data.symbol,
		});
	}
}

File: golang/pkg/database/shipyards.go
package database

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"git.adyxax.org/adyxax/spacetraders/golang/pkg/model"
)

func (db *DB) LoadShipyard(symbol string) (*model.Shipyard, error) {
	var buf []byte
	if err := db.QueryRow(`SELECT data FROM shipyards WHERE data->>'symbol' = ?;`, symbol).Scan(&buf); err != nil {
		return nil, fmt.Errorf("failed to query row: %w", err)
	}
	var shipyard model.Shipyard
	if err := json.Unmarshal(buf, &shipyard); err != nil {
		return nil, fmt.Errorf("failed to unmarshal shipyard: %w", err)
	}
	return &shipyard, nil
}

func (db *DB) SaveShipyard(shipyard *model.Shipyard) error {
	data, err := json.Marshal(shipyard)
	if err != nil {
		return fmt.Errorf("failed to marshal shipyard: %w", err)
	}
	_, err = db.Exec(
		`INSERT INTO shipyards(data, updated)
                VALUES (json(:data), :updated)
         ON CONFLICT DO UPDATE SET data = :data, updated = :updated
		                       WHERE data->>'symbol' = :symbol;`,
		sql.Named("data", data),
		sql.Named("symbol", shipyard.Symbol),
		sql.Named("updated", time.Now()),
	)
	if err != nil {
		return fmt.Errorf("failed to exec: %w", err)
	}
	return nil
}
